### ISSUE TITLE
All bgr/a textures are rendered with a wrong swizzle

### ISSUE THREAD

**Issue by elect86** [g-truc/gli#79]
I don't get.. when I retrieve the gl swizzles

`Gl.translate(texture.swizzles())`

It looks right

`{SWIZZLE_BLUE, SWIZZLE_GREEN, SWIZZLE_RED, SWIZZLE_ALPHA}`

or 

`{0x1905, 0x1904, 0x1903, 0x1906}`

But the rendering is wrong, if I comment the swizzles setting

`//gl4.glTexParameterIiv(glTarget.value, GL4.GL_TEXTURE_SWIZZLE_RGBA, glSwizzles.toArray(), 0);`

then they render properly...

And I double checked with your Gli, I get the same values for the same textures..

What is it wrong? Could it be the textures themselves?

**Comment by Groovounet** [comment]
Hi,
What version of GLI are you using? 
Can you attach you DDS?

```
		gli::gl GL;

		glActiveTexture(GL_TEXTURE0);
		glGenTextures(texture::MAX, &TextureName[0]);

		{
			gli::texture2D Texture(gli::load((getDataDirectory() + TEXTURE_DIFFUSE_DXT5_SRGB).c_str()));
			assert(!Texture.empty());
			gli::gl::format const Format = GL.translate(Texture.format());
			gli::gl::swizzles const Swizzles = GL.translate(Texture.swizzles());

			glBindTexture(GL_TEXTURE_2D, TextureName[texture::BPTC]);
			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_BASE_LEVEL, 0);
			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAX_LEVEL, GLint(Texture.levels() - 1));
			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_SWIZZLE_R, Swizzles[0]);
			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_SWIZZLE_G, Swizzles[1]);
			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_SWIZZLE_B, Swizzles[2]);
			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_SWIZZLE_A, Swizzles[3]);
			glTexStorage2D(GL_TEXTURE_2D, GLint(Texture.levels()),
				Format.Internal, GLsizei(Texture.dimensions().x), GLsizei(Texture.dimensions().y));

			for(std::size_t Level = 0; Level < Texture.levels(); ++Level)
			{
				glCompressedTexSubImage2D(GL_TEXTURE_2D, GLint(Level),
					0, 0,
					GLsizei(Texture[Level].dimensions().x), GLsizei(Texture[Level].dimensions().y),
					Format.Internal,
					GLsizei(Texture[Level].size()),
					Texture[Level].data());
			}
		}
```

**Comment by Groovounet** [comment]
I could reproduce the problem with GLI master branch. 

The problem is that both external and swizzle are BGRA which causes a double negation.

It's actually a tricky problem due to OpenGL versions differences. OpenGL ES 2.0 and OpenGL 3.2 and earlier doesn't have swizzling support. OpenGL 3.2 supports BGRA but ES 2.0 only throw extensions.

There is actually quite a few incompatibilities that I am planning to fix by adding some parameters to gli::gl. 
```
gli::gl GL(GLI_GL_CORE, 4, 5);
```
But this is work to do... \o/

**Comment by elect86** [comment]
Same also with `kueken7_r5g6b5_unorm` dds and ktx

**Comment by Groovounet** [comment]
Hi,

I have been investigating this issue and figure out a few things:
- There is still some work to do on texture classes swizzle. Doing that at the moment.
- OpenGL swizzle translation doesn't consider the texture format. It's plain translation.
To fix this _gli::gl::swizzles gl::translate(gli::swizzles const& Swizzle) const;_ need a parameter gli::format.

I probably need to think more about this. One of the annoying part is that KTX doesn't support texture swizzle (lame!). Also, GLI uses the same translation code for user translation and for KTX loading and saving. So right now, the best is to now use the swizzle translation to GL while creating an OpenGL texture.

Thanks,
Christophe

**Comment by Groovounet** [comment]
This issue is fixed for GLI 0.8 release.

Thanks for contributing,
Christophe

**Comment by elect86** [comment]
Thank you for taking care of that!
